### PR TITLE
chore: simplify native functions

### LINKED ISSUE #297: Bug: `$AWS.Lambda.Invoke` should return a promise
It does not return a promise, it should.

### PR DESCRIPTION
**Pull request by thantos**
Closes #297 

Simplify how native functions are handled, move all handling to synth time.

Depends on: https://github.com/functionless/nodejs-closure-serializer/pull/10

* [x] Replace NativeFunctionDecl with FunctionDecl AST
* [x] Replace `native.call` re-write with serialize level replacement
* [x] Errors
* [x] Validate don't create resources in closure

### CODE REVIEW

File: src/function.ts [hunk 1]
```diff
@@ -656,24 +700,42 @@ export class Function<
...
     );
   }
 }
 
+function getInvokedIntegrations(ast: FunctionDecl): IntegrationCallExpr[] {
+  const nodes: IntegrationCallExpr[] = [];
+  visitEachChild(ast, function visit(node: FunctionlessNode): FunctionlessNode {
+    if (isIntegrationCallExpr(node)) {
+      nodes.push(node);
+    }
```
Comment (sam-goodwin) [inline src/function.ts:731]: Here, a ref() could be another `function`. We will need to get its closure and recurse in. That's a change I was working on with the `global.wrapClosure` and `getClosure(func): FunctionDecl | undefined` .

File: src/function.ts [hunk 2]
```diff
@@ -604,22 +610,31 @@ export class Function<
...
+    magic?: any
   ) {
-    const func = validateFunctionlessNode(
-      isNativeFunctionOrError(propsOrFunc)
-        ? propsOrFunc
-        : isNativeFunctionOrError(funcOrNothing)
-        ? funcOrNothing
-        : undefined,
+    const func =
+      typeof propsOrFunc === "function" ? propsOrFunc : funcOrNothing;
```
Comment (sam-goodwin) [inline src/function.ts:617]: lol I did this exactly. Glad we are on the same wavelength.

File: src/function.ts [hunk 3]
```diff
@@ -950,19 +1011,28 @@ export async function serialize(
...
-              INTEGRATION_TYPE_KEYS.filter((key) => key !== "native").forEach(
-                (key) => delete copy[key]
-              );
-
-              return copy;
+              const c = integ.native?.call;
+              const call =
+                typeof c !== "undefined"
+                  ? function (...args: any[]) {
+                      return c(args, preWarmContext);
```
Comment (sam-goodwin) [inline src/function.ts:1018]: I love this bootstrap of preWarmContext

Comment (sam-goodwin) [reply src/function.ts:1018]: I think this means we don't have to rewrite ast to accept prewarm context anywhere now right?

**Comment by thantos** [comment]
https://github.com/functionless/nodejs-closure-serializer/pull/10 will fix the failing build.

File: test/function.localstack.test.ts [hunk 4]
```diff
@@ -996,6 +996,314 @@ localstackTestSuite("functionStack", (testResource, _stack, _app) => {
...
+        }
+      );
+    },
+    null
+  );
+
+  /**
+   * This should fail?
+   */
+  test.skip(
```
Comment (sam-goodwin) [inline test/function.localstack.test.ts:1150]: Amazing job writing tests ahead of time. Love the discipline and thoroughness.

**Review by sam-goodwin — APPROVED** [review]

File: src/function.ts (hunk 1 above)
Comment (thantos) [reply src/function.ts:731]: yeah, ha. this worked for the current state.

File: src/function.ts (hunk 3 above)
Comment (thantos) [reply src/function.ts:1018]: yep, no more native function specific code in compile (except for transforming the closure to ast)

File: src/function.ts (hunk 1 above)
Comment (sam-goodwin) [reply src/function.ts:731]: No I'm really happy for this change. Saves me doing it. This is in the right state now

**Comment by sam-goodwin** [comment]
118 commits? Lol. Did you branch off of async

File: tsconfig.dev.json [hunk 5]
```diff
@@ -8,7 +8,8 @@
     "inlineSources": true,
     "lib": [
       "dom",
-      "ES2019"
+      "ES2020",
+      "ES2020.Promise"
```
Comment (sam-goodwin) [inline tsconfig.dev.json:12]: What prompted this?

File: src/error-code.ts [hunk 6]
```diff
@@ -176,12 +176,12 @@ export namespace ErrorCodes {
   /**
    * Generic error message to denote errors that should not happen and are not the fault of the Functionless library consumer.
    *
-   * Please report this issue
+   * Please report this issue.
```
Comment (sam-goodwin) [inline src/error-code.ts:179]: Link to repo?

File: src/error-code.ts [hunk 7]
```diff
@@ -653,6 +652,40 @@ export namespace ErrorCodes {
     type: ErrorType.ERROR,
     title: "Unsupported use of Promises",
   };
+
+  /**
+   * Unable to fund reference out of application function.
```
Comment (sam-goodwin) [inline src/error-code.ts:657]: Find not fund ;)

Comment (thantos) [reply src/error-code.ts:657]: 1 am documentation is dangerous.

File: tsconfig.dev.json (hunk 5 above)
Comment (thantos) [reply tsconfig.dev.json:12]: Can probably switch back to be cleaner... `Promise.allSettled` was added with es2020. Didn't end up using it.

**Pull request merged**